pkg/reconciler/kubernetes/tracker_test.go

```diff
@@ -0,0 +1,331 @@
+package kubernetes
+
+import (
+	"context"
+	"github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress"
+	"io/ioutil"
+	"strings"
+
+	e "github.com/kyma-incubator/reconciler/pkg/error"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+
+	"path/filepath"
+	"testing"
+	"time"
+
+	log "github.com/kyma-incubator/reconciler/pkg/logger"
+
+	"github.com/kyma-incubator/reconciler/pkg/test"
+	"github.com/stretchr/testify/require"
+)
+
+func TestProgressTracker(t *testing.T) {
+	test.IntegrationTest(t)
+
+	logger := log.NewLogger(true)
+
+	kubeClient, err := NewKubernetesClient(test.ReadKubeconfig(t), zap.NewNop().Sugar(), nil)
+
+	require.NoError(t, err)
+
+	clientSet, err := kubeClient.Clientset()
+	require.NoError(t, err)
+
+	resources := readManifestToUnstructured(t, "all.yaml")
+	require.Len(t, resources, 5)
+
+	cleanup := func() {
+		t.Log("Cleanup test resources")
+		for _, resource := range resources {
+			deletedResource, err := kubeClient.DeleteResource(
+				resource.GetKind(), resource.GetName(), resource.GetNamespace())
+			if err != nil && !errors.IsNotFound(err) {
+				t.Fatalf("Failed to delete resource: %s", err)
+			}
+			t.Logf("Removed test resource '%s", deletedResource)
+		}
+	}
+	cleanup()       //ensure relicts from previous test runs were removed
+	defer cleanup() //cleanup after test is finished
+
+	//install test resources
+	t.Log("Creating test resources")
+	manifest := test.ReadFile(t, filepath.Join("test", "all.yaml"))
+	_, err = kubeClient.Deploy(context.TODO(), string(manifest), "")
+	require.NoError(t, err)
+
+	t.Run("Test progress tracking with timeout", func(t *testing.T) {
+		t.Skip("All resources are deployed and in ready state, what's the meaning to test timeout here?")
```

Comment:
I skipped this test for now, after clarifying, we can consider either removing or refactoring it.
- not sure if this test scenario is meaningful
- the current setup does not support this test scenario as described, the resource deployment is already finished before the test, there is no way this progress tracker can find components that are not in the ready state.

Reply:
- The motivation behind this test is to check if the `Watch` method of the `Tracker` type is reacting to a given context. The [context](https://pkg.go.dev/context) can be useful e.g. if we want to [cancel](https://pkg.go.dev/context#example-WithCancel) or [timeout](https://pkg.go.dev/context#example-WithTimeout) some operation.

- The scenario asumes that the resources will not be in `ready` state in [~1.25](https://github.com/kyma-incubator/reconciler/blob/d9ca902808702612482098048c956b8dfe6ab116/pkg/reconciler/kubernetes/progress/tracker_test.go#L76) second after the apply method was called and that the error is [propagated](https://github.com/kyma-incubator/reconciler/blob/d9ca902808702612482098048c956b8dfe6ab116/pkg/reconciler/kubernetes/progress/tracker_test.go#L79) (the `context deadline exceeded` in our case).

```
Running tool: /usr/local/bin/go test -timeout 30s -run ^TestProgressTracker$ github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress -v

=== RUN   TestProgressTracker
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:42: Cleanup test resources
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:49: Removed test resource '&{unittest-progress     Namespace}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:49: Removed test resource '&{unittest-deployment unittest-progress    Deployment}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:49: Removed test resource '&{unittest-job unittest-progress    Job}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:49: Removed test resource '&{unittest-daemonset unittest-progress    DaemonSet}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:49: Removed test resource '&{unittest-statefulset unittest-progress    StatefulSet}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:56: Creating test resources
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:60: Deployed test resource '&{unittest-progress     Namespace}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:60: Deployed test resource '&{unittest-deployment unittest-progress    Deployment}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:60: Deployed test resource '&{unittest-job unittest-progress    Job}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:60: Deployed test resource '&{unittest-daemonset unittest-progress    DaemonSet}
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:60: Deployed test resource '&{unittest-statefulset unittest-progress    StatefulSet}
=== RUN   TestProgressTracker/Test_progress_tracking_with_timeout
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:300: Register watchable Deployment 'unittest-deployment' in namespace 'unittest-progress'
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:300: Register watchable Job 'unittest-job' in namespace 'unittest-progress'
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:300: Register watchable DaemonSet 'unittest-daemonset' in namespace 'unittest-progress'
    /Users/i508150/sources/go/src/github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes/progress/tracker_test.go:300: Register watchable StatefulSet 'unittest-statefulset' in namespace 'unittest-progress'
{"level":"DEBUG","time":"2022-01-03T17:41:04.805+0100","message":"Transition of unittest-deployment to ready state is still ongoing"}
{"level":"DEBUG","time":"2022-01-03T17:41:05.782+0100","message":"Stop checking progress of resource transition to state 'ready' because parent context got closed"}
{"level":"INFO","time":"2022-01-03T17:41:05.782+0100","message":"Tracker stopped checking the progress of the following resource: Deployment [namespace:unittest-progress|name:unittest-deployment]"}
{"level":"INFO","time":"2022-01-03T17:41:05.782+0100","message":"Tracker stopped checking the progress of the following resource: Job [namespace:unittest-progress|name:unittest-job]"}
{"level":"INFO","time":"2022-01-03T17:41:05.782+0100","message":"Tracker stopped checking the progress of the following resource: DaemonSet [namespace:unittest-progress|name:unittest-daemonset]"}
{"level":"INFO","time":"2022-01-03T17:41:05.782+0100","message":"Tracker stopped checking the progress of the following resource: StatefulSet [namespace:unittest-progress|name:unittest-statefulset]"}
```